src/openlcb/NonAuthoritativeEventProducer.cxx

```diff
@@ -0,0 +1,258 @@
+            stateCallback_((event->event - eventBase_) / 2, value);
+        }
+    }
+    else
+    {
+        if (event->event >= eventBaseOn_ &&
+            event->event < (eventBaseOn_ + size_))
+        {
+            stateCallback_((event->event - eventBase_), true);
+        }
+        else if (event->event >= eventBaseOff_ &&
+                 event->event < (eventBaseOff_ + size_))
+        {
+            stateCallback_((event->event - eventBase_), false);
+        }
+    }
+}
+
+//
+// BitRangeNonauthoritativeEventP::handle_consumer_identified()
+//
+void BitRangeNonAuthoritativeEventP::handle_consumer_identified(
+                                                const EventRegistryEntry& entry,
+                                                EventReport *event,
+                                                BarrierNotifiable *done)
+{
+    done->notify();
+    if (!stateCallback_)
+    {
+        // there is nobody to notify
+        return;
+    }
+
+    bool value;
+    if (event->state == EventState::VALID)
+    {
+        value = true;
+    }
+    else if (event->state == EventState::INVALID)
+    {
+        value = false;
+    }
+    else
+    {
+        return; // nothing to learn from this message.
+    }
+
+    if (eventBaseOff_ == 0)
+    {
+        if (event->event >= eventBase_ &&
+            event->event < (eventBase_ + (size_ * 2)))
+        {
+            if ((event->event % 2) == (eventBase_ % 2))
+            {
+                stateCallback_((event->event - eventBase_) / 2, value);
+            }
+            else
+            {
+                stateCallback_((event->event - eventBase_) / 2, !value);
+            }
+        }
+    }
+    else
+    {
+        if (event->event >= eventBaseOn_ &&
+            event->event < (eventBaseOn_ + size_))
+        {
+            stateCallback_((event->event - eventBase_), value);
+        }
+        else if (event->event >= eventBaseOff_ &&
+                 event->event < (eventBaseOff_ + size_))
+        {
+            stateCallback_((event->event - eventBase_), !value);
+        }
+    }
+}
+
+//
+// BitRangeNonauthoritativeEventP::handle_identify_producer()
+//
+void BitRangeNonAuthoritativeEventP::handle_identify_global(
+                                                const EventRegistryEntry& entry,
+                                                EventReport *event,
+                                                BarrierNotifiable *done)
+{
+    if (event->dst_node && event->dst_node != node_)
+    {
+        done->notify();
+    }
+    else
+    {
+        uint64_t range;
+        if (entry.event == eventBase_)
+        {
+            range = EncodeRange(eventBase_,
+                                eventBaseOff_ == 0 ? size_ * 2 : size_);
+        }
+        else if (entry.event == eventBaseOff_ && eventBaseOff_ != 0)
+        {
+            range = EncodeRange(eventBaseOff_, size_);
+        }
+        else
+        {
+            // uninteresting event range
```

Comment:
Is this not also a problem on the top end as well?  For example, if the range ends before a power of 2 boundary, I should not respond to a request beyond the end of my range.